Make type generic

emmatyping opened the pull request.

emmatyping committed:
Make type generic

This change specifically allows type[int], but disallows str[int].

emmatyping committed:
Merge branch 'pep585' into type

gvanrossum commented:
I'll look at this tomorrow, or later -- I may be getting sick. :-(

If you rebase all tests should be passing: In test_typing.py, I changed a few assertIs into assertEqual and commented out two expected failures, and it now passes for me.

emmatyping committed:
Merge branch 'pep585' of github.com:gvanrossum/cpython into type

emmatyping committed:
Merge branch 'pep585' into type

emmatyping commented:
Rebased!

>I'll look at this tomorrow, or later -- I may be getting sick. :-(

Oh no! Sorry to hear that.

>If you rebase all tests should be passing: In test_typing.py, I changed a few assertIs into assertEqual and commented out two expected failures, and it now passes for me.

Yep things seem to be passing. I also noticed a few warnings on compile which I will submit a PR to fix.

emmatyping committed:
Merge branch 'pep585' into type

emmatyping commented:
Rebased on the pickle changes so this should be ready for review/merge.

gvanrossum commented on Lib/test/test_genericalias.py:
```diff
@@ -176,6 +176,19 @@ class L(list): ...
+        Test = t('Test', (), {})
+        self.assertTrue(isinstance(Test, type))
+        test = Test()
+        self.assertEqual(t(test), Test)
```
Maybe also check that `t(0) == int`?

gvanrossum commented on Lib/test/test_genericalias.py:
```diff
@@ -176,6 +176,19 @@ class L(list): ...
+        class MyType(type):
+            pass
+        with self.assertRaises(TypeError):
+            MyType[int]
```
Interestingly, this is different from `test_generic_subclass` (at line 96 above).

I'm okay with that, but I wonder if you're just testing an implementation accident here? Else what's the reason we want `MyType[int]` to fail, but we want `MyList[int]` to work?

emmatyping replied:
I believe `type` is special in this case. Generic containers like `list[str]` are type constructors that build `list`s. It would make sense `MyList[str]` would make `MyList`s. However, with `type`, there is only `typing.Type`, there is no other way to construct the `type` of an instance. It doesn't (AFAIK) make any sense to allow `MyType[SomeClass]`.

I can remove this test if that seems unreasonable.

gvanrossum reviewed (approved):
To be sure, I'm glad that this was all that it took to make it work!

emmatyping committed:
Add test for t(0) == int

emmatyping committed:
Merge branch 'type' of github.com:ethanhs/cpython into type

gvanrossum reviewed (approved):
Eh, I think it's fine. I'll just land this, I'm very happy that we got this working!

emmatyping commented:
>I'm very happy that we got this working!

Me too! On to the next thing :)

EDIT: But before that https://github.com/gvanrossum/cpython/pull/10
